### PR TITLE
Fix: AnVIL indexer doesn't follow downstream links from files to files (#4761)

### LINKED ISSUE #4761: AnVIL indexer doesn't follow downstream links from files to files
https://github.com/DataBiosphere/azul/blob/e19471e3935bb784acd983a7ae932d8004a5a1be/src/azul/plugins/repository/tdr_anvil/__init__.py#L262

Uses the plural `'files'` where the proper key is singular `'file'`. Since the container is a `defaultdict`, an empty set is returned instead instead of raising a `KeyError`. This causes us to skip the step where we follow downstream links from files to other files. There are other bugs in this step that are not being detected as a result.

The impact appears to be low. When I reindexed my personal deployment (with the same source config as `anvilbox`) with a fix in place, the number of indexed files remained constant at 174. The fix I used was:

```
diff --git a/src/azul/plugins/repository/tdr_anvil/__init__.py b/src/azul/plugins/repository/tdr_anvil/__init__.py
index 9ee680e7..fc5c5910 100644
--- a/src/azul/plugins/repository/tdr_anvil/__init__.py
+++ b/src/azul/plugins/repository/tdr_anvil/__init__.py
@@ -229,7 +229,7 @@ class Plugin(TDRPlugin):
         return bundle_entity
 
     def _consolidate_by_type(self, entities: Keys) -> MutableKeysByType:
-        result = defaultdict(set)
+        result = {entity_type: set() for entity_type in self.indexed_columns_by_entity_type}
         for e in entities:
             result[e.entity_type].add(e.key)
         return result
@@ -259,7 +259,7 @@ class Plugin(TDRPlugin):
                            ) -> Links:
         return set.union(
             self._downstream_from_biosamples(source, entities['biosample']),
-            self._downstream_from_files(source, entities['files'])
+            self._downstream_from_files(source, entities['file'])
         )
 
     def _upstream_from_biosamples(self,
@@ -407,8 +407,8 @@ class Plugin(TDRPlugin):
         rows = self._run_sql(f'''
             WITH activities AS (
                 SELECT
-                    ala.alignmentactivity_id,
-                    'alignmentactivity',
+                    ala.alignmentactivity_id AS activity_id,
+                    'alignmentactivity' AS activity_table,
                     ala.used_file_id,
                     ala.generated_file_id
                 FROM {backtick(self._full_table_name(source, 'alignmentactivity'))} AS ala
@@ -433,7 +433,7 @@ class Plugin(TDRPlugin):
                             KeyReference(key=file_id, entity_type='file')
                             for file_id in row['generated_file_id']
                         ],
-                        activity=KeyReference(key=row['actvity_id'], entity_type=row['activity_table']))
+                        activity=KeyReference(key=row['activity_id'], entity_type=row['activity_table']))
             for row in rows
         }
 
@@ -442,6 +442,8 @@ class Plugin(TDRPlugin):
                            entity_type: EntityType,
                            keys: AbstractSet[Key],
                            ) -> MutableJSONs:
+        if not keys:
+            return []
         table_name = self._full_table_name(source, entity_type)
         columns = set.union(
             self.common_indexed_columns,
```

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by nadove-ucsc**
Connected issues: #4761 

### Author

- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue        <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via Zenhub 
- [x] PR description links to connected issues
- [x] Added `partial` label to PR                                   <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues      <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (requirements, before every review)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased PR branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations         <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or all connected issues are labeled `no demo`</sub>
- [x] Rebased and squashed PR branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label    <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment                          <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `sandbox` deployment      <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev`                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment                         <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build log for anomalies in `anvilbox` deployment     <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `anvilbox`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `anvilbox`                                 <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox`                            <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [x] Pushed merge commit to GitLab `dev`                           <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev`                      <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build log for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is* 
pushed determines this checklist item.

### Operator (reindex) 

- [x] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Deleted unreferenced indices in `anvildev`                    <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev`                                 <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev`            <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev`       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment                       <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment                  <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by dsotirho-ucsc — DISMISSED** [review]
Approved.

File: src/azul/plugins/repository/tdr_anvil/__init__.py [hunk 1]
```diff
@@ -442,6 +442,8 @@ def _retrieve_entities(self,
                            entity_type: EntityType,
                            keys: AbstractSet[Key],
                            ) -> MutableJSONs:
+        if not keys:
+            return []
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr_anvil/__init__.py]: No bailouts, even if it makes the diff bigger. Most diff tools, including GH are pretty good at de-emphasizing whitespace changes.

File: src/azul/plugins/repository/tdr_anvil/__init__.py [hunk 2]
```diff
@@ -407,8 +407,8 @@ def _downstream_from_files(self,
         rows = self._run_sql(f'''
             WITH activities AS (
                 SELECT
-                    ala.alignmentactivity_id,
-                    'alignmentactivity',
+                    ala.alignmentactivity_id AS activity_id,
+                    'alignmentactivity' AS activity_table,
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr_anvil/__init__.py]: This seems to be an unrelated change, maybe a fix for a bug that was shadowed by #4761? If so, that should go into a separate commit.

File: src/azul/plugins/repository/tdr_anvil/__init__.py [hunk 3]
```diff
@@ -433,7 +433,7 @@ def _downstream_from_files(self,
                             KeyReference(key=file_id, entity_type='file')
                             for file_id in row['generated_file_id']
                         ],
-                        activity=KeyReference(key=row['actvity_id'], entity_type=row['activity_table']))
+                        activity=KeyReference(key=row['activity_id'], entity_type=row['activity_table']))
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr_anvil/__init__.py]: Same here?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by dsotirho-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
@danielsotirhos, when you are operator, you must not unassign yourself with open checklist items.

**Comment by dsotirho-ucsc** [comment]
I previously failed to mark the checkbox `Build passes on GitLab anvildev` for the merge commit.
Marking the checkbox now, proof of build passing is here: https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/pipelines/2498

**Pull request merged**